### ISSUE TITLE
Recovering metric depth when mixing datasets during training

### ISSUE THREAD

**Issue by anna-monica** [nianticlabs/monodepth2#241]
Hello,
I'm a big fan of your work, and I've had the chance to play around with your model and see its magnificent results!

I have tried to train/finetune the model myself with custom datasets and would like further information on how to recover metric depth. I have checked previous issues and know that if we train the model in stereo mode, we can recover metric depth using the camera baseline, and if we train in mono mode we will need ground truth depths in order to apply median scaling. 
I have however some doubts regarding scaling in some particular training scenarios:

1- If I use your stereo (or mono+stereo) trained KITTI model and then fine-tune it on my custom dataset (3k pairs), what would be the scaling factor? would it be the KITTI baseline or the custom dataset's baseline?
2- Same question when mixing datasets with different camera intrinsics and extrinsics during stereo training for better generalization, how would we recover the scaling factor in that case?
3- If I'm using mono-training mode on a dataset that has no ground truth depth maps, what would be the minimum required knowledge to apply median scaling?
4- If I use your stereo trained KITTI model to inference depth on "wild images" is there any way to go from disparity to metric depth, or would we have to stick with relative depth values?

I would really appreciate any help you can offer or resources you can recommend as I am still new in the computer vision field.
Thank you in advance!

**Comment by mrharicot** [comment]
Hi,
Thanks for the kind words!

1. If you train only with your custom dataset at the end then the scaling factor should be based on your own dataset's baseline, not kitti.
2. Mixing multiple stereo datasets gets tricky. We currently hardcode the baseline to 0.1 units in the stereo case (in the dataloader). In your case you would want to have that baseline to be dependent on which dataset the sample is from, and make sure the relative scale between them is correct.
3. Median scaling is just used for evaluation. In the case of lack of ground truth you can try to find objects of known size in some frames and compute the relative scale from that.
4. You can simply run the kitti evaluation to get the average scale (it prints it at the end) and use that as your "best guess". However if the test distribution is far away from kitti you can't really expect your model to produce anything very accurate.

I hope this helps!

**Comment by anna-monica** [comment]
Hi,
Thank you for your prompt response!

Your answers were very helpful! I would need some further details concerning questions 2 and 3 if that's okay.

> 2. Mixing multiple stereo datasets gets tricky. We currently hardcode the baseline to 0.1 units in the stereo case (in the dataloader). In your case you would want to have that baseline to be dependent on which dataset the sample is from

I don't really see how (on what basis) would I change the hardcoded baseline during training. Is it not possible to let the network learn with the 0.1 baseline (and different intrinsics) for both Datasets and then recover the metric scale for each dataset according to its own "real" baseline?

> and make sure the relative scale between them is correct.

I don't seem to understand which relative scale we are referring to in that case... can you possibly further explain the process to follow? 

> 3. Median scaling is just used for evaluation. In the case of lack of ground truth you can try to find objects of known size in some frames and compute the relative scale from that.

Just to make sure I understood correctly: if I happen to know for example that a pedestrian has a height H of 1.7 meters and that on the image he appears to have a height h of 20 pixels, I can recover his true depth using the focal distance f (depth/f = h/H) and then use this value to recover the scaling factor (real_depth/monodepth_depth) and apply it to all my depth maps?

Thanks again!

**Comment by mdfirman** [comment]
> Is it not possible to let the network learn with the 0.1 baseline (and different intrinsics) for both Datasets and then recover the metric scale for each dataset according to its own "real" baseline?

Yes – I think in theory it is possible. It might need a more careful scaling of weights or outputs though. Overall the output of the network (including the scaled sigmoid) has received relatively little attention in academic works and could do with some more thorough exploration.

> and make sure the relative scale between them is correct.

If I'm understanding this correctly, the point here is that if you train from one stereo dataset with a baseline of 0.2m mixed with a different stereo dataset with a baseline of 0.6m, you can't just naively mix pairs from the two datasets. The different baselines (or at least the ratios between the baselines) need to be taken into account too.

For your question on 3: Yes, absolutely. You are looking for a single 'multiplier' factor with which to rescale all your depth outputs with. If you are able to capture your own data, a simpler way might be to point the camera at an object you know is say 5m away. If the network predicts a depth of 1.25 for this object, you know you should multiply all your depth outputs by 4 to bring them into 'real world' scale range.

**Comment by mrharicot** [comment]
Closing this issue for now.
